app/models/publication.rb

```diff
@@ -70,10 +69,32 @@ def is_public?
     is_published? && !is_embargoed? && !is_yanked?
   end
 
+  def collaborators
+    authors.map(&:user) + copyright_holders.map(&:user)
+  end
+
   def has_collaborator?(user)
-    authors.any?{|a| a.user == user} || \
-    copyright_holders.any?{|ch| ch.user == user} || \
-    editors.any?{|e| e.user == user}
+    collaborators.include? user
+  end
+
+  def has_read_permission?(user)
+    has_collaborator?(user) || collaborators.any? do |collaborator|
+      collaborator.list_owners.any? do |list_owner|
+        list = list_owner.list
+        list.has_publication_group?(publication_group) && list.has_member?(user)
+      end
+    end
```

Comment:
OK I got rid of most of them... might still be a performance issue if we start using Accounts groups as list owners, but for now should be fine...